### DISCUSSION TITLE
Why does prisma reset the dev database we are applying migrations to?

### DISCUSSION THREAD

**Discussion by sk8Guerra** [prisma/prisma#8914] [Q&A]
I made the init migration, of the whole database in the development environment and every table was created. Then I filled some tables with data. I added some new relations between those previously created tables and when I run the command to apply new migrations Prisma says: 

![Screen Shot 2021-08-24 at 22 05 12](https://user-images.githubusercontent.com/18681676/130724824-1a22989c-789d-48df-b2fd-8c3906e55ac7.png)

Why is that? shouldn't it apply the new migrations without touching the data? Will that also happen in production? All the tables will be empty after applying migrations.

Thanks for helping in advance!

**Comment by ryands17** [comment]
@sk8Guerra 👋 

Can you share what new models/fields did you add that gives this message?

**Reply by sk8Guerra** [reply]
@ryands17 👋🏽😃

```prisma
model role {
  id             Int           @id @default(autoincrement())
  name           String        @unique
  users          users[]
  institutionid  Int?
  institution    institution?  @relation(fields: [institutionid], references: [id])
}
```

Those are the fields I added to the `role` model. Basically a new 1-to-many relation between `institution` and `role`.

Thanks for the help!

**Comment by janpio** [comment]
Can you also share what was output _before_ what you screenshotted? There should be some information on _why_ there is drift detected. Also, which version of Prisma are you using?

**Reply by sk8Guerra** [reply]
I'm using **primsa: 2.30.0** and **@prisma/client: 2.22.1**. Give me some mins, I'll take the screenshot again.

**Reply by sk8Guerra** [reply]
Well, well, well... This time didn't force me to reset the database. 🤔

![Screen Shot 2021-08-25 at 10 27 09](https://user-images.githubusercontent.com/18681676/130828893-467c666b-6d0c-4bfe-b0be-8c595d8b594b.png)

**Reply by sk8Guerra** [reply]
I suppose it is because this time I added the new fields with a **?** sign. And when I check the modified table, every record has a null value. And the data is there.

<img width="426" alt="Screen Shot 2021-08-25 at 10 33 10" src="https://user-images.githubusercontent.com/18681676/130829805-daf75b94-a6f0-45d1-aeb6-9c406cac61f7.png">

**Reply by sk8Guerra** [reply]
When is a drift detected then?

**Reply by ryands17** [reply]
You cannot add a field as required when there's already data in the table as that would be a null constraint violation. Whenever you want to add new fields to an already existing database, you need to add it as optional `?` first.

> When is a drift detected then?

The main error that you posted in the question was not actually a drift but a constraint violation that couldn't be solved without data deletion. A [drift](https://www.prisma.io/docs/guides/database/developing-with-prisma-migrate/troubleshooting-development#schema-drift) occurs when your schema is out of sync with your migrations.

**Reply by sk8Guerra** [reply]
@ryands17, great explanation, thank you so much! 🙏🏼

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
